### PR TITLE
fishing for #1225

### PR DESCRIPTION
**Pull request by ssiloti**
Changes triggered by the hunt for the root cause of #1225

### CODE REVIEW

File: src/disk_io_thread.cpp [hunk 1]
```diff
@@ -2319,22 +2326,25 @@ namespace libtorrent
...
 			if (next_locked_block < num_locked_blocks
 				&& locked_blocks[next_locked_block] == i)
 			{
 				++next_locked_block;
 				TORRENT_PIECE_ASSERT(pe->blocks[i].buf, pe);
-				TORRENT_PIECE_ASSERT(ph->offset == i * block_size, pe);
-				ph->offset += int(iov.iov_len);
+				TORRENT_PIECE_ASSERT(offset == i * block_size, pe);
+				offset += int(iov.iov_len);
 				ph->h.update({pe->blocks[i].buf, iov.iov_len});
```
Comment (arvidn) [inline src/disk_io_thread.cpp]: there better only be a single thread calling ``update()`` here. that's guaranteed, right?

Comment (ssiloti) [inline src/disk_io_thread.cpp]: Yes, I verified that the hasher is not accessed concurrently.

File: src/disk_io_thread.cpp [hunk 2]
```diff
@@ -2084,22 +2086,27 @@ namespace libtorrent
 		DLOG("kick_hasher: %d - %d (piece: %d offset: %d)\n"
```
Comment (arvidn) [inline src/disk_io_thread.cpp]: we set ``hashing = 1`` right above here. as far as I can tell, we don't release the mutex between checking ``hashing`` and setting it, meaning that there really only should be one thread getting here at a time, no?

File: src/disk_io_thread.cpp (hunk 1 above)
Comment (arvidn) [inline src/disk_io_thread.cpp]: I suppose some other thread may still read ``offset``, which justifies this fix

File: src/disk_io_thread.cpp (hunk 2 above)
Comment (ssiloti) [inline src/disk_io_thread.cpp]: There are other places that offset is read without checking `hashing`, in particular `try_flush_hashed`.

File: src/disk_io_thread.cpp [hunk 3]
```diff
@@ -2411,6 +2421,8 @@ namespace libtorrent
 
 		l.lock();
 
+		ph->offset = offset;
```
Comment (arvidn) [inline src/disk_io_thread.cpp]: should there be a TORRENT_ASSERT(ph->offset == offset); here and above too?

**Comment by arvidn** [comment]
awesome. nice catch!

**Comment by arvidn** [comment]
is this a problem in RC_1_1 as well?

File: src/disk_io_thread.cpp (hunk 2 above)
Comment (ssiloti) [inline src/disk_io_thread.cpp]: And to answer you question, yes, only one thread should be kicking the hasher at a time.

**Comment by ssiloti** [comment]
I'm not sure it actually is a problem. Unless the compiler is doing something truly evil (to be fair, it wouldn't be the first time) this shouldn't cause a problem on any remotely common ISA. Reliance on 32-bit loads/stores being atomic is very common practice.

**Comment by arvidn** [comment]
I think the fact that it's a race condition (or rather that the compiler can assume there's no concurrent access) opens up all kinds of possibilities that won't necessarily have intuitive behavior.

File: src/disk_io_thread.cpp [hunk 4]
```diff
@@ -2084,22 +2086,31 @@ namespace libtorrent
 		DLOG("kick_hasher: %d - %d (piece: %d offset: %d)\n"
 			, cursor, end, int(pe->piece), ph->offset);
 
+		// save a local copy of offset to avoid concurrent access
+		int offset = ph->offset;
+#if TORRENT_USE_ASSERTS
+		int old_offset = offset;
```
Comment (arvidn) [inline src/disk_io_thread.cpp]: oh, right. my first pass reading this, I assumed ``offset`` was immutable. then I realized it wasn't, but I was still thinking that the assert I proposed would work.. oh well

**Comment by arvidn** [comment]
lgtm

**Comment by ssiloti** [comment]
Alright, if you want it on RC_1_1 then I'll rebase and re-submit.

**Pull request closed without merging**